### PR TITLE
Gaussian checkpoint writer

### LINKED ISSUE #617: Tool for writing Gaussian formatted checkpoint file
A number of external tools, prominently GDMA https://gitlab.com/anthonyjs/gdma, work with input as formatted checkpoint file from Gaussian. The format is described for example here: http://wild.life.nctu.edu.tw/~jsyu/compchem/g09/g09ur/f_formchk.htm

To ensure interoperability with these external tools, we should make a converter from our data to *.fchk.

### PR DESCRIPTION
**Pull request by rubengerritsen**
A new tool that converts a .orb file to a .fchk (formatted gaussian checkpoint) file, fix #617. 

The file specification for the .fchk file on the Gaussian website is very limited, the email from the gaussian people helped a bit, but also didn't contain everything I needed. What I did to get the order, multipliers etc. right is run a dft calculation with psi4, run exactly the same calculation with xtp and compare the votca generated .fchk file with the .fchk output file of psi4. Hence if the psi4 implementation is correct, this one is as well. I couldn't think of any other way to verify the correctness of the implementation. If someone has a good idea please let me know.

Also I still need to add a unit test.

### CODE REVIEW

**Comment by rubengerritsen** [comment]
@votca-bot format

**Comment by junghans** [comment]
@votca-bot copyright

File: src/libxtp/gaussianwriter.cc [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+ * Gaussian uses a minus sign to indicate spherical shells.
+ */
+Index GaussianWriter::toGaussianL(L l) const {
+  switch (l) {
+    case L::S:
+      return 0;
+    case L::P:
+      return 1;
+    default:
+      return -1 * static_cast<Index>(l);
```
Comment (JensWehner) [inline src/libxtp/gaussianwriter.cc:22]: why do you have to switch for S and P?

File: src/libxtp/gaussianwriter.cc [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+            0,1,-1,2,-2,3,-3,4,-4, //g
+            0,1,-1,2,-2,3,-3,4,-4,5,-5, // h
+            0,1,-1,2,-2,3,-3,4,-4,5,-5,6,-6 // i
+  }};
+  // clang-format on
+  OrbReorder reorder(gaussianOrder, multipliers, true);
+  Eigen::MatrixXd density = orbitals.DensityMatrixGroundState();
+  reorder.reorderOperator(density, orbitals.SetupDftBasis());
+
+  votca::tools::EigenIO_MatrixMarket::WriteMatrix("density.mm", density);
```
Comment (JensWehner) [inline src/libxtp/gaussianwriter.cc:87]: I think this should go?

File: src/libxtp/gaussianwriter.cc [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+
+  std::ofstream outFile(basename + ".fchk");
+  if (outFile.is_open()) {
+    int temp_int;
+    // job description
+    outFile << basename << ", fchk created by VOTCA-XTP\n";
+    outFile << "DUMMY_TYPE    DUMMY_METHOD    " << orbitals.getDFTbasisName()
+            << "\n";
+
+    // clang-format off
```
Comment (JensWehner) [inline src/libxtp/gaussianwriter.cc:124]: ?

File: src/libxtp/tools/orb2fchk.cc [hunk 4]
```diff
@@ -0,0 +1,41 @@
+#include "orb2fchk.h"
```
Comment (JensWehner) [inline src/libxtp/tools/orb2fchk.cc:1]: copyright

File: src/libxtp/gaussianwriter.cc [hunk 5]
```diff
@@ -0,0 +1,307 @@
+#include "votca/xtp/gaussianwriter.h"
```
Comment (JensWehner) [inline src/libxtp/gaussianwriter.cc:1]: copyright

**Review by JensWehner — CHANGES_REQUESTED** [review]
Small nitpicks and add a test pls. I am not fully sure how to best test this. We probably should have a function in tools which compares two files for equality.

File: src/libxtp/gaussianwriter.cc (hunk 2 above)
Comment (rubengerritsen) [reply src/libxtp/gaussianwriter.cc:87]: Yes

File: src/libxtp/gaussianwriter.cc (hunk 1 above)
Comment (rubengerritsen) [reply src/libxtp/gaussianwriter.cc:22]: Because a -1 for the p shell translates to an sp shell in gaussian and -0 doesn't make much sense.

Comment (JensWehner) [reply src/libxtp/gaussianwriter.cc:22]: okay

File: src/libxtp/tools/orb2fchk.cc [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+// Local VOTCA includes
+#include "votca/xtp/gaussianwriter.h"
+#include "votca/xtp/orbitals.h"
+#include <votca/tools/constants.h>
+
+namespace votca {
+namespace xtp {
+
+void Orb2Fchk::ParseOptions(const tools::Property& options) {
```
Comment (JensWehner) [inline src/libxtp/tools/orb2fchk.cc:33]:
```suggestion
void Orb2Fchk::ParseOptions(const tools::Property&) {
```

File: src/libxtp/tools/orb2fchk.h [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+class Orb2Fchk final : public QMTool {
+ public:
+  Orb2Fchk() = default;
+
+  ~Orb2Fchk() = default;
+
+  std::string Identify() { return "mol2orb"; }
+
+ protected:
+  void ParseOptions(const tools::Property& user_options);
```
Comment (JensWehner) [inline src/libxtp/tools/orb2fchk.h:42]:
```suggestion
  void ParseOptions(const tools::Property&);
```

**Comment by JensWehner** [comment]
@votca-bot changelog: implement fchk writer

**Comment by JensWehner** [comment]
@votca-bot format

**Comment by rubengerritsen** [comment]
@votca-bot format

**Review by JensWehner — APPROVED** [review]

**Pull request merged**